### DISCUSSION TITLE
Encryption issue on Tx retry (frame 802.15.4-2015)

### DISCUSSION THREAD

**Discussion by EricSTMicro** [openthread/openthread#8496] [Q&A] [answered]
Hello,

I would like to be sure on the right manner to how to handle retry frames.

On our platform, we are currently using the OT lib on a commit from last July (SHA1: 32ac6ddf).

We have the OT_RADIO_CAPS_TRANSMIT_SEC capability set, so the OpenThread stack may or may not dedicate the frame's encryption to the platform. That is working well.

The problem is when a Tx retry is done on that frame (usually because no ACK).

On IEEE802.15.4-2003/2006 frames, I don't have any problem, but on IEEE802.15.4-2015 frames, some times the frame is already encrypted --> OT layer takes the same frame already encrypted by platform, some times it gives the same raw frame to encrypt again.
The issue is that the boolean _mIsSecurityProcessed_ is always set to 0 on a retry on frame v2015, thus, platform concludes it has to encrypt the frame, but in fact it has already be done (by platform, on last failed Tx). So we end up encrypting an already encrypted frame, which is problematic, instead of just sending the frame as it is.
On contrary, the boolean _mIsHeaderUpdated_ (in case of IE present) is correctly updated between a 1st try and the retries.
The boolean _mIsARetx_ is also correctly set on a retry.

In radio.h, we have the following information: _"If the platform updates the header, it must also set this flag before passing the frame back from the `otPlatRadioTxDone()` callback."_ So this is done correctly, I tried to do the same for _mIsSecurityProcessed_, but it didn't work. 

Question is: what is the correct way to handle this? What point am I missing there? I feel like _mIsHeaderUpdated_ has something do to with my issue.

Best Regards,
Eric

**Comment by abtink** [comment]
Hi @EricSTMicro 

My expectation is that `mIsSecurityProcessed` would be set whenever OT stack does process the security on the frame (independent of other flags). If it is not the case it may be an issue/bug somewhere in the code. 
We have this in the code (which should call `SetIsSecurityProcessed ()` and set the flag on processing security):
https://github.com/openthread/openthread/blob/27be812bad663e56a4feec7adf25af0f8a65e7c1/src/core/mac/mac_frame.cpp#L1176-L1206

Can you maybe provide an example of frame that is already encrypted with `mIsSecurityProcessed` being false?
What are your set of `OT_RADIO_CAPS`?

**Reply by EricSTMicro** [reply]
Hello abtink, 
thank you for your answer.

My OT_RADIO_CAPS: (OT_RADIO_CAPS_ACK_TIMEOUT | OT_RADIO_CAPS_ENERGY_SCAN | OT_RADIO_CAPS_CSMA_BACKOFF | OT_RADIO_CAPS_TRANSMIT_SEC)

I included a wireshark capture (networkkey: 00112233445566778899aabbccddeef0) with a MTD doing a child update request (to initiate CSL). I reset the FTD just before to make the retry happen.

I my code, I managed to isolate the problem, which happens when CSL is enabled.
I also printed the boolean.

What I did to have it work is when frame is a retry and CSL is present I don't encrypt it again. It works but you can see mIsSecurityProcessed is never set. Yet on the capture you can see the frame is well encrypted.

Correspond to frame #29 in wireshark log:
```
mIsARetx= 0 | mIsHeaderUpdated= 0 | mIsSecurityProcessed= 0 | mCslPresent= 1
[TXa:x8c|00911000|..|0014]      c11d -> 1ead    (14)
[0000018976] [REGION UNDEF] [I] Mac-----------: Frame tx attempt 1/16 failed, error:NoAck, len:106, seqnum:140, type:Data, src:33334444d041c11d, dst:11112222d0411ead, sec:yes
```
Correspond to frame #30 to #32 in wireshark log:
```
mIsARetx= 1 | mIsHeaderUpdated= 1 | mIsSecurityProcessed= 0 | mCslPresent = 1
[TXa:x8c|00111000|..|0014]      c11d -> 1ead    (14)
[0000018986] [REGION UNDEF] [I] Mac-----------: Frame tx attempt 2/16 failed, error:NoAck, len:106, seqnum:140, type:Data, src:33334444d041c11d, dst:11112222d0411ead, sec:yes
mIsARetx= 1 | mIsHeaderUpdated= 1 | mIsSecurityProcessed= 0 | mCslPresent= 1

 [TXa:x8c|00111000|..|0014]      c11d -> 1ead    (14)
[0000018996] [REGION UNDEF] [I] Mac-----------: Frame tx attempt 3/16 failed, error:NoAck, len:106, seqnum:140, type:Data, src:33334444d041c11d, dst:11112222d0411ead, sec:yes
mIsARetx= 1 | mIsHeaderUpdated= 1 | mIsSecurityProcessed= 0 | mCslPresent= 1

[TXa:x8c|00111000|..|0014]      c11d -> 1ead    (14)
[0000019005] [REGION UNDEF] [I] Mac-----------: Frame tx attempt 4/16 failed, error:NoAck, len:106, seqnum:140, type:Data, src:33334444d041c11d, dst:11112222d0411ead, sec:yes
```

**Reply by EricSTMicro** [reply]
Better with the .cap file :)
[log_ch_upd_req_retry_already_encrypted.zip](https://github.com/openthread/openthread/files/10184073/log_ch_upd_req_retry_already_encrypted.zip)

**Comment by abtink** [comment, marked as answer]
Thanks @EricSTMicro 

I have a theory/guess on this:

On the first tx attempt where `mIsARetx= 0 | mIsHeaderUpdated= 0 | mIsSecurityProcessed= 0 | mCslPresent= 1` I expect the radio platform does process tx security on the passed-in frame. Do you set the `mIsSecurityProcessed` to `true` after doing so on the `otRadioFrame` instance?

I think this may be the reason. Here is the flow of frame:
- Since we don't use `OT_RADIO_CAPS_TRANSMIT_RETRIES` in the configs then OT stack will handle frame tx retries. 
- After the first tx attempt failure (`NoAck)`, `SubMac` layer will pass back the same `otRadioFrame` instance back to radio platform for next tx attempt. 
- In this case OT stack will only mark it as a retx (set `mIsARetx`), but leave all other flags `mIsHeaderUpdated` and `mIsSecurityProcessed` as they are are (from previous attempt).

**Reply by EricSTMicro** [reply]
Hello @abtink,

I was indeed returning the `mIsSecurityProcessed` value (in my last trials only, I wasn't doing it originally as I didn't see any changes), but I wasn't always correctly setting it after a platform encryption... my bad. But I wasn't sure if it was necessary to return this value to OT layer, so I have my answer now, thank you. 

It is now working as expected.

One last question: on a frame retry, the value `mIsARetx` is set, are we supposed to treat the frame differently compared to this value?

**Reply by abtink** [reply]
Glad to hear it is working :)

On `mIsARetx` I think radio platform can ignore it. It is used within OT stack. When radio provides `OT_RADIO_CAPS_TRANSMIT_SEC` then `mIsHeaderUpdated` and `mIsSecurityProcessed` are the two main flags.
